Clarity improvements for porting guide content (#179)
Quansight-Labs/free-threaded-compatibility: docs/porting.md

# Porting Python Packages to Support Free-Threading

This document discusses porting an existing Python package to support free-threading Python.

## Current status (as of early 2025)

Many Python packages, particularly packages relying on C
extension modules, do not consider multithreaded use or make strong
assumptions about the GIL providing sequential consistency in multithreaded
contexts. These packages will:

- fail to produce deterministic results on the free-threaded build
- may, if there are C extensions involved, crash the interpreter in multithreaded use in ways that are impossible on the
    GIL-enabled build

> Because of the GIL, attempting to parallelize many workflows using the Python
> [threading](https://docs.python.org/3/library/threading.html) module will not
> produce any speedups, so thread safety issues that are possible even with the

Author: ```suggestion
Attempting to parallelize many workflows using the Python
[threading](https://docs.python.org/3/library/threading.html) module will not
produce any speedups on the GIL-enabled build, so thread safety issues that are possible even with the
```

protect you from introducing thread safety issues due to [race
conditions](https://en.wikipedia.org/wiki/Race_condition). It is possible to
write algorithms that depend on the precise timing of threads completing
work. It is up to you as a user of multithreaded parallelism to ensure that
simultaneous reads and writes to the same Python variable are impossible.

If you're maintaining a library, we suggest documenting what multithreaded

> workflows are supported and to document to what the library guards against

Author: ```suggestion
workflows are supported and to document to what extent the library guards against
```

races. If a variable is protected by a lock, you should document that, as it may
impact multithreaded scaling. Similarly, if a mutable data structure does not

> offer locking, you should also document that explicitly.

Author: ```suggestion
ensure state is synchronized if an object is shared between threads, you should also document that explicitly.
```

introducing the possibility of data races in native extensions and race
conditions that are impossible when the GIL is enabled.

> We suggest focusing on safety and multithreaded scaling over single-threaded
> performance. For example, if adding a lock to a global cache would harm
> multithreaded scaling, and turning off the cache implies a small performance
> hit, consider doing the simpler thing and disabling the cache in the
> free-threaded build. Single-threaded performance can always be improved later,
> once you've established free-threaded support and hopefully improved test
> coverage for multithreaded workflows.

Reviewer: ```suggestion
We suggest focusing on safety and multithreaded scaling before single-threaded
performance. 

Here's an example of this approach. If adding a lock to a global cache would harm
multithreaded scaling, and turning off the cache implies a small performance
hit, consider doing the simpler thing and disabling the cache in the
free-threaded build. 

Single-threaded performance can always be improved later,
once you've established free-threaded support and hopefully improved test
coverage for multithreaded workflows.
```

Of course this introduces a scaling bottleneck when `SafeCounter` instances are
concurrently updated. It's possible to implement more optimized locking
strategies, but doing so requires knowledge of the problem.

> ## Third-party libraries

Reviewer: ```suggestion
### Third-party libraries
```

Reviewer: I think you should consider downgrading "Dealing with mutable global state" to H3.
